lib/StackAnalysis/SegregateDirectStackAccesses.cpp

```diff
@@ -0,0 +1,152 @@
+/// \file SegregateDirectStackAccesses.cpp
+/// \brief Segregate direct stack accesses from all other memory accesses
+/// through alias information. This provides a way to say that stack accesses do
+/// not interfere with any other memory access.
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/MDBuilder.h"
+#include "llvm/IR/PatternMatch.h"
+
+#include "revng/StackAnalysis/SegregateDirectStackAccesses.h"
+#include "revng/Support/Assert.h"
+
+using namespace llvm;
+using SDSAP = SegregateDirectStackAccessesPass;
+
+PreservedAnalyses SDSAP::run(Function &F, FunctionAnalysisManager &FAM) {
+  Context = &(F.getContext());
+
+  // Get the result of the GCBI analysis
+  auto &M = *F.getParent();
+  const auto &MAMProxy = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F);
+  GCBI = MAMProxy.getCachedResult<GeneratedCodeBasicInfoAnalysis>(M);
+  if (!GCBI)
+    GCBI = &(FAM.getResult<GeneratedCodeBasicInfoAnalysis>(F));
+  revng_assert(GCBI != nullptr);
+
+  // Populate the two buckets with all load and store instruction of the
+  // function, properly segregated.
+  segregateAccesses(F);
+
+  // Adorn IR with the alias information collected before.
+  decorateStackAccesses();
+
+  return PreservedAnalyses::none();
+}
+
+void SDSAP::segregateAccesses(Function &F) {
+  using namespace PatternMatch;
+  Value *LoadSP = nullptr;
+  bool Found = false;
+
+  for (BasicBlock &BB : F) {
+    for (Instruction &I : BB) {
+      if (auto *LI = dyn_cast<LoadInst>(&I)) {
+        if (GCBI->isSPReg(skipCasts(LI->getPointerOperand()))) {
+          LoadSP = LI;
+          Found = true;
+          break;
+        }
+      }
+    }
+
+    if (Found)
+      break;
+  }
+
+  // Context: inttoptr instructions basically inhibits all optimizations. In
+  // particular, when an integer is inttoptr'd twice with different destination
+  // type, alias analysis messes up. Hence, we need to ensure that no inttoptr
+  // exists when operating on a instruction that directly accesses the stack.
+  // Note that this problem will be addressed by opaque pointers in the future.
+  IRBuilder<> Builder(&F.getEntryBlock().front());
+  auto *I8PtrTy = Builder.getInt8PtrTy();
+  Value *SPI8Ptr = Builder
+                     .CreateLoad(I8PtrTy,
+                                 ConstantExpr::getBitCast(GCBI->spReg(),
+                                                          I8PtrTy
+                                                            ->getPointerTo()));
```

Comment:
* [ ] Ugly indentation, split in multiple variables.
* [ ] You're creating an instruction at the very beginning of the basic block. The first basic block should start with `alloca`s. It's a convention many passes rely upon. Find the first-non `alloca` instruction and insert your stuff there.

Reply:
You're making an assumption you're not asserting.
Option 1: you assert the first instruction is not an `alloca`.
Option 2: you don't make the assumption and make sure you're skipping over `alloca`s.

Reply:
Checked twice, and apparently there are a few cases in which, at this point, I still have some `alloca`s. I'm going to insert the load of `SP` after the `alloca`s.